pageserver/src/layered_repository/layer_map.rs

```diff
@@ -136,7 +147,17 @@ impl LayerMap {
     /// Insert an on-disk layer
     ///
     pub fn insert_historic(&mut self, layer: Arc<dyn Layer>) {
-        let segentry = self.segs.entry(layer.get_seg_tag()).or_default();
+        let tag = layer.get_seg_tag();
+        let segentry = self.segs.entry(tag).or_default();
+
+        let mut e = self.last_seg.entry(tag.rel).or_insert(RelishSize {
+            lsn: layer.get_end_lsn(),
+            last_segno: tag.segno,
+        });
+        if e.last_segno < tag.segno {
+            e.last_segno = tag.segno;
+            e.lsn = layer.get_end_lsn();
+        }
```

Comment:
Ok, I agree with you that less code -> less bugs. And may be really it is better not to commit this optimization right now. But in future I think lazy loading of layer map is required in any case.
So I tried to address the problem with maintaining last segment cache in load_historical().
It is really non trivial. So I will be please if can verify that new implementation is correct:
```
    pub fn insert_historic(&mut self, layer: Arc<dyn Layer>) {
        let tag = layer.get_seg_tag();
        let segentry = self.segs.entry(tag).or_default();
		let lsn = layer.get_end_lsn();
		let is_dropped = layer.is_dropped();
		let last_segno = tag.segno;
        let mut e = self.last_seg.entry(tag.rel).or_insert(RelishSize {
            lsn,
            last_segno,
			is_dropped,
        });
		if (is_dropped && e.last_segno > last_segno && (e.lsn <= lsn || e.is_dropped)) ||
		   (!is_dropped && e.last_segno < last_segno && (e.lsn <= lsn || !e.is_dropped))
		{
            e.last_segno = last_segno;
            e.lsn = lsn;
			e.is_dropped = is_dropped;
        }
        segentry.insert_historic(layer);

        NUM_ONDISK_LAYERS.inc();
    }
```

Reply:
> Ok, I agree with you that less code -> less bugs. And may be really it is better not to commit this optimization right now. > But in future I think lazy loading of layer map is required in any case.

I agree we need to do something about init_current_logical_size. Like calculate it lazily. Caching the last segment won't help if you have a lot of small tables instead of one big one. But let's put that in the backlog and revisit later.

> So I tried to address the problem with maintaining last segment cache in load_historical().
> It is really non trivial. So I will be please if can verify that new implementation is correct:

I don't quite understand what's going on with `is_dropped`. I suppose it's correct but I'm not sure..